src/arch/io/network.cc

```diff
@@ -621,6 +621,430 @@ void linux_tcp_conn_t::on_event(int /* events */) {
     event_watcher->stop_watching_for_errors();
 }
 
+/* tls_conn_wrapper_t wraps a TLS connection and safely releases it in
+exceptional cases if it is not reclaimed before execution leaves its scope. */
+class tls_conn_wrapper_t {
+public:
+    tls_conn_wrapper_t(SSL *_conn) {
```

Comment:
Actually even better: Just call `SSL_new` directly in the `tls_conn_wrapper_t` constructor rather than passing it in.

There are two additional things you'll need to do:
1. According to https://www.openssl.org/docs/manmaster/ssl/SSL_new.html `SSL_new` can return `nullptr`. You'll need to check for that and throw a `connect_failed_exc_t`.
2. Make sure that if a connection is thrown, the value of the `conn` field is still initialized to `nullptr`. Then in the `~tls_conn_wrapper_t` destructor, check if `conn` is `nullptr` and only call `SSL_free` if it's not. (throwing the exception from the constructor will immediately run the destructor, so it has to work even in the error case)